### PR TITLE
Feature: Method to replace an edge

### PR DESCRIPTION
**Pull request by steven-causalens**
## Description and Motivation
This PR adds the `CausalGraph.replace_edge` method that allows you to replace an existing edge with a new one.

Among other things, this allows you to swap the direction of an edge.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (non-breaking change which improves functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactor (moving code around, general refactoring improvements)
- [ ] Documentation (documentation)

## PR Checklist:

- [ ] I have reviewed my own PR? (review the PR as you are reviewing someone else's PR)
- [ ] I have implemented all requirements? (see JIRA, project documentation)
- [ ] I am affecting someone else's work? If so: I have added those people as reviewers?
- [ ] I have added comments to all the bits that are hard to follow
- [ ] At a bare minimum, I tested this manually
- [ ] I have added unit test(s)
- [ ] Is this a bugfix? If so have I added a regression test?
- [ ] Does this PR satisfy the [one PR, one concern](https://medium.com/@fagnerbrack/one-pull-request-one-concern-e84a27dfe9f1) rule?
- [ ] If needed, documentation has been added/updated?
- [ ] I have updated the appropriate changelog with a line for my changes
- [ ] If this PR breaks reproducibility, have I added a note in the changelog explaining the break in reproducibility
      and its extent?

### CODE REVIEW

File: tests/test_causal_graph.py
Comment (andrew-causalens) [inline tests/test_causal_graph.py:1]: Worth testing new method with the `TimeSeriesCausalGraph` subclass.

File: cai_causal_graph/causal_graph.py [hunk 1]
```diff
@@ -1279,6 +1279,55 @@ def remove_edge(self, /, source: NodeLike, destination: NodeLike, *, edge_type:
...
+        :param new_destination: Destination of the new edge.
+        :param edge_type: The edge type of the new edge. If `None` (default), the same edge type as the original edge
+            will be used.
+        :param meta: Optional meta to be used for the new edge. If `None` (default)`, the meta of the original edge
+            will be used.
+        """
+        # ensure the new edge does not exist and the original one does
+        assert self.edge_exists(
+            source=source, destination=destination
+        ), f'The provided edge ({source}, {destination}) to be replaced does not exist'
```
Comment (ilya-causalens) [inline cai_causal_graph/causal_graph.py:1314]: Should raise `cai_causal_graph.exceptions.CausalGraphErrors.EdgeDoesNotExistError`, OR simply don't check and let `get_edge` raise this.

File: cai_causal_graph/causal_graph.py [hunk 2]
```diff
@@ -1279,6 +1279,55 @@ def remove_edge(self, /, source: NodeLike, destination: NodeLike, *, edge_type:
...
+        assert not self.edge_exists(
+            source=new_source, destination=new_destination
+        ), f'Cannot create a new edge ({source}, {destination}) as it already exists'
+
```
Comment (ilya-causalens) [inline cai_causal_graph/causal_graph.py:1318]: Raise `cai_causal_graph.exceptions.CausalGraphErrors.EdgeExistsError`

**Review by ilya-causalens — APPROVED** [review]
LGTM, but need better errors. Also, as Andrew said - should test TSCG.

File: tests/test_time_series_graph.py [hunk 3]
```diff
@@ -273,6 +273,22 @@ def test_replace_node(self):
         # check 'X1 lag(n=1)' does not exist
         self.assertFalse(ts_cg.node_exists('X1 lag(n=1)'))
 
+    def test_replace_edge(self):
```
Comment (andrew-causalens) [inline tests/test_time_series_graph.py:276]: Can you add an error case where you try to replace with a new edge that flows backwards in time? Just confirm it errors.

**Review by andrew-causalens — APPROVED** [review]
LGTM. Just a request for a slight extension to unit test.

**Pull request merged**